File: Programmers/CPPSolutions/조이스틱/Joystick.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(string name)
{
    int answer = 0;
    int first = -1, last = -1;
    for (int i = 0; i < name.length(); i++)
    {
        if (name[i] != 'A')
        { // A가 아닐 때 문자 변환 카운트
            answer += min((int)('Z' - name[i]) + 1, (int)(name[i] - 'A'));
        }
        else
        { // A면 first, last 인덱스 구하기
            if (first == -1)
            {
                first = i - 1;
            }
            if (last == -1 && first != -1)
                for (int j = i + 1; j < name.length(); j++)
                {
                    if (name[j] != 'A')
                    {
                        last = j;
                        break;
                    }
                }
        }
    }

    if (first == -1 && last == -1) // last,first 둘 다 없는 경우
        answer += name.length() - 1;
    else
    {
        if (first != -1 && last == -1) // last만 있는 경우
            return answer += first;
        else if (first == -1 && last != -1) // first만 있는 경우
            return answer += name.length() - last;
        else
        { // last, first 둘 다 있는 경우
            int dis = min(first * 2 + (name.length() - last), (name.length() - last) * 2 + first);
            return answer += min((int)name.length() - 1, dis);
        }
    }
}

int main()
{
    int answer = solution("BBBAAAAABAAAAAAAAAAA");
    printf("%d", answer);
    return 0;
}

File: Swift/Programmers/ProgrammersSolutions/ProgrammersSolutions/조이슽/Joystick.swift
import Foundation

final class Joystick {
  func solution(_ name:String) -> Int {
    let stringAlphabets = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    var alphabetDict = [String: Int]()
    stringAlphabets.enumerated().forEach({ alphabetDict[String($1)] = $0 })
    let nameArray = name.map({ String($0) })
    var answer = 0
    var minimumMove = nameArray.count - 1
    

    for alphabet in nameArray {
      if alphabet != "A" {
        let z = alphabetDict["Z"] ?? 25
        let a = alphabetDict["A"] ?? 0
        let alphabetIndex = alphabetDict[String(alphabet)] ?? 0
        answer += min(z - alphabetIndex + 1, alphabetIndex - a)
      }
    }
    
    
    for i in 0 ..< nameArray.count {
      var next = i + 1
      while next < nameArray.count && nameArray[next] == "A" {
        next += 1
      }
      let leftOrRight = min(i + (i + nameArray.count - next), 2 * (nameArray.count - next) + i)
      minimumMove = min(minimumMove, leftOrRight)
    }
    
    answer += minimumMove
    
    return answer
  }
}
